scipy/sparse/csgraph/_flow.pyx

```diff
@@ -680,3 +709,881 @@ cdef ITYPE_t[:] _dinic(
+
+    Returns
+    -------
+    res : MinimumCostFlowResult
+        A minimum cost flow represented by a ``MinimumCostFlowResult``
+        which includes the cost of the flow in ``flow_cost`` and
+        the flow graph in ``flow``.
+
+    Raises
+    ------
+    TypeError:
+        if the input graph is not in CSR format.
+
+    ValueError:
+        If any of the following conditions is met:
+          - the capacity values, vertex demands or edge costs are not integers
+          - an edge cost or vertex demand are infinite in absolute value
+          - an edge has negative capacity
+          - the sum of demands is not zero
+          - there doesn't exist a flow which satisfies all vertex demands.
+
+    Notes
+    -----
+    This solves the minimum cost flow problem [1]_ on a given directed weighted graph:
+    a flow associates to every edge a value, also called a flow, less than the
+    capacity of the edge, and for provided cost of the edge so that for every
+    vertex (apart from the source and the sink vertices), the difference between
+    the total incoming flow and the total outgoing flow is equal to the demand
+    of that vertex. The value of a flow is the sum of the flow on all edges and
+    the minimum cost flow problem consists of finding a flow whose cost (which is
+    defined as the scalar product of the edges' flows and costs) is minimal.
+
+    To solve the problem, we use Network Simplex [2]_.
+    The time complexity of the former :math:`O(|V|^2\,|E|\,log(|V|\,C))`.
+
+    The minimum cost flow problem is usually defined with real valued capacities,
+    edge costs and vertex demands but we require that all capacities are integral.
+    When dealing with rational capacities, or capacities belonging to
+    :math:`x\mathbb{Q}` for some fixed :math:`x \in \mathbb{R}`, it is possible
+    to reduce the problem to the integral case by scaling all capacities
+    accordingly.
+
+    Solving a minimum cost flow problem has many practical uses such as
+    in assignment problem, covers and matching in bipartite graphs [3]_.
+
+    We have ported the Python implementation of NetworkX [4]_ to Cython.
+
+    References
+    ----------
+
+    .. [1] https://en.wikipedia.org/wiki/Minimum-cost_flow_problem#Definition
+    .. [2] Orlin, J.B. A polynomial time primal network simplex algorithm
+           for minimum cost flows. Mathematical Programming 78,
+           109-129 (1997).
+    .. [3] https://en.wikipedia.org/wiki/Network_simplex_algorithm#Applications
+    .. [4] A Hagberg, D Schult, P Swart, Exploring Network Structure, Dynamics,
+           and Function using NetworkX in Proceedings of the 7th Python in
+           Science conference (SciPy 2008), G Varoquaux, T Vaught, J Millman
+           (Eds.), pp. 11-15
+
+    Examples
+    --------
+    >>> from scipy.sparse import csr_matrix
+    >>> from scipy.sparse.csgraph import minimum_cost_flow
+    >>> A = csr_matrix([[0, 4, 10, 0],
```

Comment:
Here's a simple example where I don't get what I expect:

```
A= csr_matrix([[0, 1], [0, 0]])
demand = [-1, 1]
cost = [2]
result = minimum_cost_flow(A, demand, cost)
```
Here, I would expect a flow whose cost is 2. Instead I get an

```python
IndexError: Out of bounds on buffer access (axis 0)
```

Reply:
Note that e.g. `maximum_flow(A, 0, 1)` works as expected.

Reply:
If I change `cost` to `cost = [2, 2]`, I do get a result. There's only one edge though, so something isn't quite right.

Reply:
Here's a more general example; the one corresponding to the regression test in `test_flow`:

```python
n = 100
indices = np.arange(1, n)
indptr = np.array(list(range(n)) + [n - 1])
data = np.ones(n - 1, dtype=np.int32)
graph = csr_matrix((data, indices, indptr), shape=(n, n))
res = maximum_flow(graph, 0, n - 1)
assert res.flow_value == 1
expected_flow = graph - graph.transpose()
minimum_cost_flow(graph, [-1] + [0] * (n-2) + [1], [1] * (n-1))
```

Here, I would also expect a non-trivial flow but get a

```python
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
IndexError: Out of bounds on buffer access (axis 0)

Exception ignored in: '_cython_magic_16fa31bcfbe023d295063980d3b33003._find_cycle'
Traceback (most recent call last):
  File "/tmp/ipykernel_16739/59043073.py", line 1, in <cell line: 1>
IndexError: Out of bounds on buffer access (axis 0)

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [10], in <cell line: 1>()
----> 1 minimum_cost_flow(graph, [-1] + [0] * (n-2) + [1], [1] * (n-1))

File _cython_magic_16fa31bcfbe023d295063980d3b33003.pyx:857, in _cython_magic_16fa31bcfbe023d295063980d3b33003.minimum_cost_flow()

ValueError: no flow satisfies all node demands
```

Notably, if I change `n` to `n = 100_000`, it doesn't seem to complete at all.

Reply:
Here's a bit more info for debugging: `minimum_cost_flow(csr_matrix([[0, 0], [1, 0]]), [1, -1], [2])` becomes

```python
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
IndexError: Out of bounds on buffer access (axis 0)

Exception ignored in: '_cython_magic_16fa31bcfbe023d295063980d3b33003._find_cycle'
Traceback (most recent call last):
  File "/tmp/ipykernel_16739/2476308831.py", line 1, in <cell line: 1>
IndexError: Out of bounds on buffer access (axis 0)

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
IndexError: Out of bounds on buffer access (axis 0)

Exception ignored in: '_cython_magic_16fa31bcfbe023d295063980d3b33003._find_leaving_edge'
Traceback (most recent call last):
  File "/tmp/ipykernel_16739/2476308831.py", line 1, in <cell line: 1>
IndexError: Out of bounds on buffer access (axis 0)

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
IndexError: Out of bounds on buffer access (axis 0)

Exception ignored in: '_cython_magic_16fa31bcfbe023d295063980d3b33003._augment_flow'
Traceback (most recent call last):
  File "/tmp/ipykernel_16739/2476308831.py", line 1, in <cell line: 1>
IndexError: Out of bounds on buffer access (axis 0)

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
IndexError: Out of bounds on buffer access (axis 0)

Exception ignored in: '_cython_magic_16fa31bcfbe023d295063980d3b33003._index'
Traceback (most recent call last):
  File "/tmp/ipykernel_16739/2476308831.py", line 1, in <cell line: 1>
IndexError: Out of bounds on buffer access (axis 0)
```

Reply:
If it helps, this example does work with NetworkX, which might help in finding out where there's a discrepancy:

```python
In [13]: G = nx.DiGraph()
In [14]: G.add_node('a', demand=-1)
In [15]: G.add_node('b', demand=1)
In [16]: G.add_edge('a', 'b', weight=2, capacity=1)
In [17]: nx.min_cost_flow(G)
Out[17]: {'a': {'b': 1}, 'b': {}}
In [18]: nx.min_cost_flow_cost(G)
Out[18]: 2
```

Reply:
**Code**

```python
from scipy.sparse import csr_matrix
from scipy.sparse.csgraph import minimum_cost_flow, maximum_flow
import numpy as np

A = csr_matrix([[0, 1], [0, 0]])
demand = [-1, 1]
cost = [2]
result = minimum_cost_flow(A, demand, cost)

print("Case #1")
print(result)

def case_2(n):
    indices = np.arange(1, n)
    indptr = np.array(list(range(n)) + [n - 1])
    data = np.ones(n - 1, dtype=np.int32)
    graph = csr_matrix((data, indices, indptr), shape=(n, n))
    res = maximum_flow(graph, 0, n - 1)
    assert res.flow_value == 1
    return minimum_cost_flow(graph, [-1] + [0] * (n-2) + [1], [1] * (n-1))

print("Case #2: n = 100")
print(case_2(100))

print("Case #3: n = 1000")
print(case_2(1000))
```

**Output**

```bash
Case #1
MinimumCostFlowResult with a cost of 2
Case #2: n = 100
MinimumCostFlowResult with a cost of 99
Case #3: n = 1000
MinimumCostFlowResult with a cost of 999
```

Bottomline - I don't get any error and all the outputs are getting printed. Not sure if they are correct or not. As far as I can figure out the time complexity of the algorithm is O(V^2 * E * log(VC)). Here V = n, E = n - 1 ~ n, C = 1, so O(n^3 * log(n)). For n = 100_000, the number of steps in the algorithm would be O(10^15 * 16.6). I don't expect it to complete anyways for such large values.

Reply:
Sounds good. Still not sure what could have caused it; Cython or C compiler bug maybe?

Did my noise cause us to miss the 1.9 freeze, @rgommers? Sorry about that if so; it's nicer to, well, get things out than not get them out.

Reply:
Yes indeed, but no worries about that - I really appreciate the detailed testing/review. The time to get potential algorithmic issues surfaced and fixed is before merging; post-merge typically it's much harder.

Reply:
It went out of my mind the previous Saturday. 😅. Will do it tomorrow for sure.

Reply:
Done.

Reply:
Thanks!

Reply:
Only Windows CI fails for your example so probably a windows specific problem. Someone with access to Windows should be able to fix it (or at least make an attempt towards fixing it).

Reply:
Based on the paths (`/tmp` etc.), I probably ran the example on WSL. Maybe depends on the version of Cython or another dependency, or something like that?

Reply:
I have mac and linux based systems. In both as far as I remember I receive correct outputs. I don’t have any Windows based systems so can’t say for sure. It might also be related to MSVC or any other Windows specific things which is very normal in Windows from my experience.

Reply:
A bit of a stretch, but I suppose the Cython update that came out the day after our discussion didn't help? (Can we rerun CI?)

It does seem like this would be quite the pain to debug, but it would also be wonderful to get this functionality into SciPy. If this is indeed a Cython/MSVC bug, perhaps someone from @cython might be interested in figuring out what's going on?

Reply:
>  (Can we rerun CI?)

triggered now

agreed that it'd be great to get this in

Reply:
Windows tests still fail.